cdap-app-templates/cdap-program-report/cdap-program-report-base/src/main/java/co/cask/cdap/report/proto/FilterCodec.java

```diff
@@ -22,14 +22,17 @@
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
 
 import java.lang.reflect.Type;
 import javax.annotation.Nullable;
 
 /**
- * JSON deserializer for {@link Filter}
+ * JSON codec for {@link Filter}
  */
-public class FilterDeserializer implements JsonDeserializer<Filter> {
+public class FilterCodec implements JsonSerializer<Filter>, JsonDeserializer<Filter> {
```

Comment:
This deserializer is way too complex and error prone (the existing logic, not the stuff you've added).

We should avoid custom serialization/deserialization when possible, especially for proto classes as they are supposed to represent request/response objects. Having this separate codec is brittle and error prone, which is how this bug came up in the first place. Proto objects really shouldn't be interfaces or abstract classes. They shouldn't use generics that aren't actually generic. ValueFilter is only ever with strings, it should not be generic, it should just use string. RangeFilter is only ever int or long, it should not be generic, it should just use longs.

This codec has to understand Filter and all its implementations in order to be correct, which couples everything together. It is much simpler just to have a concrete class that contains nullable whitelist, blacklist, and range fields.

Reply:
After poking around some more, it looks like only ValueFilter is ever used, and only ever on namespace. Is the use of the other filters yet to be implemented or am I missing something?

Reply:
the range filter is used for time range fields such as start, end, running time and duration of runs

Reply:
yes, that makes sense, its true the existing code is unnecessarily complex.

Reply:
maybe my IDE doesn't have everything indexed then, I only see the filter getter used in ReportGenerationSpark.ReportSparkHandler.getNamespaceFilterIfExists

Reply:
its used in `co.cask.cdap.report.ReportGenerationHelper.scala` class